### PR TITLE
Add a initial implementation of the Builder pattern

### PR DESCRIPTION
**Pull request by TheFakeMontyOnTheRun**
Not the best example (since it's my first non-trivial Kotlin code) and might not highlight the best features of Kotlin, but it's quite faithful to the ideas presented in the GOF book - which is not to be taken without a pinch of salt, since that was a (ancient) C++ book.

~TheFakeMontyOnTheRun

### CODE REVIEW

File: designpatterns/gangoffour/src/main/kotlin/com/felipecosta/designpatterns/creational/Builder.kt [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        room1.setSide(RoomSide.SOUTH, builder.makeMazeWall())
+        room1.setSide(RoomSide.WEST, builder.makeMazeWall())
+        room2.setSide(RoomSide.NORTH, builder.makeMazeWall())
+        room2.setSide(RoomSide.EAST, builder.makeMazeWall())
+        room2.setSide(RoomSide.SOUTH, builder.makeMazeWall())
+        room2.setSide(RoomSide.WEST, door)
+    }
+}
+
+class BuilderDoor(val outRoom: BuilderRoom, val inRoom: BuilderRoom)
```
Comment (felipehjcosta) [inline designpatterns/gangoffour/src/main/kotlin/com/felipecosta/designpatterns/creational/Builder.kt:38]: Using a prefix for all of this classes seems strange, they all look like a `Builder` object.
I suggest creating a new package `builder` and moving it into inside.
I will create separated packages for the others patterns as well. :)

File: designpatterns/gangoffour/src/main/kotlin/com/felipecosta/designpatterns/creational/Builder.kt [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    abstract fun makeMazeWall(): BuilderWall;
+    abstract fun makeDoor(outRoom: BuilderRoom, inRoom: BuilderRoom): BuilderDoor;
+
+    fun getResult(): BuilderMaze {
+        return mazeBeingBuilt;
+    }
+}
+
+open class MazeDirector(val builder: MazeBuilder) {
+    fun Construct() {
```
Comment (felipehjcosta) [inline designpatterns/gangoffour/src/main/kotlin/com/felipecosta/designpatterns/creational/Builder.kt:19]: What do you think of using `camelCase` here?

Comment (TheFakeMontyOnTheRun) [inline designpatterns/gangoffour/src/main/kotlin/com/felipecosta/designpatterns/creational/Builder.kt:19]: Sorry about that! Coding too much C++ and reading from a C++ book can get confusing at times!
I will fix this ASAP.

File: designpatterns/gangoffour/src/main/kotlin/com/felipecosta/designpatterns/creational/Builder.kt (hunk 1 above)
Comment (TheFakeMontyOnTheRun) [inline designpatterns/gangoffour/src/main/kotlin/com/felipecosta/designpatterns/creational/Builder.kt:38]: I will try that - but wouldn't be confusing for the reader?

Comment (felipehjcosta) [inline designpatterns/gangoffour/src/main/kotlin/com/felipecosta/designpatterns/creational/Builder.kt:38]: Consistency only. :)

**Pull request closed without merging**